### PR TITLE
Include third parties library imports in exported package.json

### PR DESCRIPTION
**Pull request by AyloSrd**
PR for [Issue #113](https://github.com/solidjs/solid-playground/issues/113). 
Currently it is possible to import third party js libraries in any playground.
These imports however, are not included in the package.json file when the current playground is exported as a zip folder, thus requiring the developer to manually `npm install` each single library.

This PR adresses this issues by :
1. Leveraging _rollup_'s feature of listing the imports for any bundle.
2. Extracting package's name and version either form the import url path, or from Skypack's response.
3. Creating a separate event for the communication between _rollup_ and the `<Header />` component, in order to avoid bloating the API at any bundle creation.

### CODE REVIEW

File: playground/utils/exportFiles.tsx [hunk 1]
```diff
@@ -68,17 +51,108 @@ const indexHTML = (tabs: Tab[]) => dedent`
...
+  function getLatestVersion(pkgName: string): Promise<string[]> {
+    return fetch(`${CDN_URL}/${pkgName}`)
+      .then(res => res.text())
+      .then(data => {
+        const pkgFound = '/*\n * Skypack CDN - ' + pkgName + '@'
+        const withPkgError = '/*\n * [Package Error] "' + pkgName + '@v'
+
+        if (data.startsWith(pkgFound)) return [pkgName, data.split(pkgFound)[1].split('\n')[0]]
+        // sometmes skypack cannot compile certains packages, but will still give us the vesion
+        if (data.startsWith(withPkgError)) return [pkgName, data.split(withPkgError)[1].split('"')[0]]
```
Comment (AyloSrd) [inline playground/utils/exportFiles.tsx:95]: The [Package Error] scenario, which happens when Skypack cannot compile a package to the `esm` format if I am not wrong (e.g try with `@babel/standalone`), can slow down this process, as, when Skypack cannot build a package, the response comes considerably slower (like 1 sec).

As an alternative we could send a request to **UNPKG**, which redirects immediatly the url to the last version, making it possible to extract the version from the new url string.

However, I preferd stick to Skypack for two reasons :
1.  You can never know if Skypack and **UNPKG** ave the same reference for a package's last version. It's better if we have a single source of truth.
2. The user will see while coding the playground that the package cannot be imported, so they probably won't download it yet.

File: playground/components/header.tsx [hunk 2]
```diff
@@ -32,6 +37,14 @@ export const Header: ParentComponent<{ fork?: () => void; share: () => Promise<s
     setShowMenu(false);
   }
 
+  onMount(() => {
+    props.compiler.addEventListener('message', ({ data }) => {
+      const { event, imports } = data;
+
+      if (event === 'IMPORTS') exportToZip(context.tabs()!, imports)
```
Comment (AyloSrd) [inline playground/components/header.tsx:44]: no need to listen to errors here, as it happens already in the REPL, and we would have a double `console.error`

**Comment by modderme123** [comment]
Somehow this slipped through my email a second time, sorry about that. Adding a second event listener is clever. 

I'm currently going through this to clean it up, and then I'll merge

**Comment by AyloSrd** [comment]
> Somehow this slipped through my email a second time, sorry about that. 

No problem at all.

> Adding a second event listener is clever!

Thanks; I thougt it would be more efficient than fishing the information down in the REPL component, and then pass it up to the parent and down again to the Header at every build in rollup. At least like this we get the imports only when we ask for them.

> I'm currently going through this to clean it up, and then I'll merge

Looking forward!

File: playground/utils/exportFiles.tsx [hunk 3]
```diff
@@ -68,17 +56,103 @@ const indexHTML = (tabs: Tab[]) => dedent`
 </html>
 `;
 
+/**
+ * This function extracts the name and version of an imported
+ * package, returning an array [name, version], in case the
+ * import si written as name@version, or otherwhise as [name, '']
+ */
+ function extractNameAndVersionFromImportPath(importPath: string): string[] {
+    const withoutCDN = importPath.startsWith(CDN_URL.domain)
```
Comment (AyloSrd) [inline playground/utils/exportFiles.tsx:65]: adapted to the new CDN

File: playground/utils/exportFiles.tsx [hunk 4]
```diff
@@ -68,17 +56,103 @@ const indexHTML = (tabs: Tab[]) => dedent`
...
+  function getLatestVersion(pkgName: string): Promise<string[]> {
+    return fetch(MAKE_CDN_URL(pkgName))
+      .then(res =>
+        extractNameAndVersionFromImportPath(res.url)[1].length ?
+          extractNameAndVersionFromImportPath(res.url)
+        :
+          [pkgName, 'latest']
+      )
+      .catch(() => [pkgName, 'latest'])
+  }
```
Comment (AyloSrd) [inline playground/utils/exportFiles.tsx:100]: adapted to the new CDN.
This is actually faster than the function used with Skypack

File: playground/pages/edit.tsx [hunk 5]
```diff
@@ -162,9 +162,20 @@ export const Edit = (props: { horizontal: boolean }) => {
...
 
+  compiler.addEventListener('message', ({ data }) => {
+    const { event, imports, error } = data;
+
+    if (error) return console.error(error);
+
+    if (event === 'IMPORTS') {
+      console.log(imports);
+    }
+  });
```
Comment (modderme123) [inline playground/pages/edit.tsx:173]: What is this for?

**Comment by modderme123** [comment]
Just reviewed this again and it looks great! Just want to understand what that extra log is doing (to see if we can remove it), and then I'll merge

File: playground/pages/edit.tsx (hunk 5 above)
Comment (AyloSrd) [reply playground/pages/edit.tsx:173]: oups I believe I must have put it there at the beginning to test it and forgot to remove it

**Comment by AyloSrd** [comment]
> Just reviewed this again and it looks great! Just want to understand what that extra log is doing (to see if we can remove it), and then I'll merge

Thank you! 
It's a log I have put at the very beginning to test, and I have forgot to remove it. There is no need for it

**Pull request merged**